File: src/desktop/modern/src/application/services/positioning/arrows/orchestration/arrow_positioning_orchestrator.py
"""
Arrow Positioning Orchestrator

Coordinates microservices to provide the same interface as the monolith.
Uses dependency injection to compose positioning pipeline.
"""

import logging
from typing import Tuple

from core.interfaces.positioning_services import (
    IArrowAdjustmentCalculator,
    IArrowCoordinateSystemService,
    IArrowLocationCalculator,
    IArrowPositioningOrchestrator,
    IArrowRotationCalculator,
)
from domain.models.arrow_data import ArrowData
from domain.models.motion_models import MotionData
from domain.models.pictograph_data import PictographData
from presentation.components.pictograph.graphics_items.arrow_item import ArrowItem

logger = logging.getLogger(__name__)


class ArrowPositioningOrchestrator(IArrowPositioningOrchestrator):
    """Orchestrates microservices to handle arrow positioning."""

    def __init__(
        self,
        location_calculator: IArrowLocationCalculator,
        rotation_calculator: IArrowRotationCalculator,
        adjustment_calculator: IArrowAdjustmentCalculator,
        coordinate_system: IArrowCoordinateSystemService,
    ):
        self.location_calculator = location_calculator
        self.rotation_calculator = rotation_calculator
        self.adjustment_calculator = adjustment_calculator
        self.coordinate_system = coordinate_system

        self.mirror_conditions = {
            "anti": {"cw": True, "ccw": False},
            "other": {"cw": False, "ccw": True},
        }

    def calculate_arrow_position(
        self,
        arrow_data: ArrowData,
        pictograph_data: PictographData,
        motion_data: MotionData,
    ) -> Tuple[float, float, float]:
        """Calculate arrow position using microservices pipeline."""
        motion = motion_data or self._get_motion_from_pictograph(
            arrow_data, pictograph_data
        )

        if not motion:
            logger.warning(
                f"No motion data for {arrow_data.color}, returning center position"
            )
            center = self.coordinate_system.get_scene_center()
            return center.x(), center.y(), 0.0

        location = self.location_calculator.calculate_location(motion, pictograph_data)
        initial_position = self.coordinate_system.get_initial_position(motion, location)

        initial_position = self._ensure_valid_position(initial_position)

        rotation = self.rotation_calculator.calculate_rotation(motion, location)
        adjustment = self.adjustment_calculator.calculate_adjustment(
            arrow_data, pictograph_data
        )

        adjustment_x, adjustment_y = self._extract_adjustment_values(adjustment)

        final_x = initial_position.x + adjustment_x
        final_y = initial_position.y + adjustment_y

        return final_x, final_y, rotation

    def calculate_all_arrow_positions(
        self, pictograph_data: PictographData
    ) -> PictographData:
        """Calculate positions for all arrows in the pictograph."""
        updated_pictograph = pictograph_data

        for color, arrow_data in pictograph_data.arrows.items():
            motion_data = pictograph_data.motions.get(color)

            if arrow_data.is_visible and motion_data:
                x, y, rotation = self.calculate_arrow_position(
                    arrow_data, pictograph_data, motion_data
                )

                updated_pictograph = updated_pictograph.update_arrow(
                    color, position_x=x, position_y=y, rotation_angle=rotation
                )

        return updated_pictograph

    def should_mirror_arrow(
        self, arrow_data: ArrowData, pictograph_data: "PictographData" = None
    ) -> bool:
        """Determine if arrow should be mirrored."""
        motion = None
        if pictograph_data and pictograph_data.motions:
            motion = pictograph_data.motions.get(arrow_data.color)

        if not motion:
            return False

        motion_type = motion.motion_type.value.lower()
        prop_rot_dir = motion.prop_rot_dir.value.lower()

        if motion_type == "anti":
            return self.mirror_conditions["anti"].get(prop_rot_dir, False)
        else:
            return self.mirror_conditions["other"].get(prop_rot_dir, False)

    def apply_mirror_transform(
        self, arrow_item: ArrowItem, should_mirror: bool
    ) -> None:
        """Apply mirror transformation."""
        try:
            from PyQt6.QtGui import QTransform

            center_x = arrow_item.boundingRect().center().x()
            center_y = arrow_item.boundingRect().center().y()

            transform = QTransform()
            transform.translate(center_x, center_y)
            transform.scale(-1 if should_mirror else 1, 1)
            transform.translate(-center_x, -center_y)

            arrow_item.setTransform(transform)
        except (ImportError, AttributeError):
            pass

    def _get_motion_from_pictograph(
        self, arrow_data: ArrowData, pictograph_data: PictographData
    ) -> MotionData:
        """Extract motion data from pictograph data."""
        if (
            not pictograph_data
            or not hasattr(pictograph_data, "motions")
            or not pictograph_data.motions
        ):
            return None
        return pictograph_data.motions.get(arrow_data.color)

    def _ensure_valid_position(self, initial_position):
        """Ensure position object has valid x and y attributes."""
        if hasattr(initial_position, "x") and hasattr(initial_position, "y"):
            return initial_position

        from core.types.geometry import Point

        return Point(475.0, 475.0)

    def _extract_adjustment_values(self, adjustment) -> Tuple[float, float]:
        """Extract x and y values from adjustment object."""
        if isinstance(adjustment, (int, float)):
            return adjustment, adjustment

        if hasattr(adjustment, "x") and hasattr(adjustment, "y"):
            x_attr = adjustment.x
            y_attr = adjustment.y
            adjustment_x = x_attr() if callable(x_attr) else x_attr
            adjustment_y = y_attr() if callable(y_attr) else y_attr
            return adjustment_x, adjustment_y

        return 0.0, 0.0

File: web/src/lib/shared/pictograph/arrow/orchestration/services/implementations/ArrowPositioningOrchestrator.ts
/**
 * Arrow Positioning Orchestrator
 *
 * Main implementation of the arrow positioning pipeline.
 * Coordinates with other services to compute final arrow positions.
 */

import type {
    ArrowPlacementData,
    IArrowAdjustmentCalculator,
    IArrowLocationCalculator,
    IArrowPositioningOrchestrator,
    MotionData,
    PictographData,
} from "$shared";
import { TYPES } from "$shared";
import { inject, injectable } from "inversify";
import type { IArrowRotationCalculator } from "../../../positioning/calculation/services/contracts";
import type { IArrowCoordinateTransformer, IArrowDataProcessor, IArrowGridCoordinateService } from "../contracts";

@injectable()
export class ArrowPositioningOrchestrator implements IArrowPositioningOrchestrator {
  private locationCalculator: IArrowLocationCalculator;
  private rotationCalculator: IArrowRotationCalculator;
  private adjustmentCalculator: IArrowAdjustmentCalculator;
  private coordinateSystem: IArrowGridCoordinateService;

  private coordinateTransformer: IArrowCoordinateTransformer;
  private dataProcessor: IArrowDataProcessor;

  constructor(
    @inject(TYPES.IArrowLocationCalculator)
    locationCalculator: IArrowLocationCalculator,
    @inject(TYPES.IArrowRotationCalculator)
    rotationCalculator: IArrowRotationCalculator,
    @inject(TYPES.IArrowAdjustmentCalculator)
    adjustmentCalculator: IArrowAdjustmentCalculator,
    @inject(TYPES.IArrowGridCoordinateService)
    coordinateSystem: IArrowGridCoordinateService,
    @inject(TYPES.IArrowCoordinateTransformer)
    coordinateTransformer: IArrowCoordinateTransformer,
    @inject(TYPES.IArrowDataProcessor)
    dataProcessor: IArrowDataProcessor
  ) {
    this.locationCalculator = locationCalculator;
    this.rotationCalculator = rotationCalculator;
    this.adjustmentCalculator = adjustmentCalculator;
    this.coordinateSystem = coordinateSystem;
    this.coordinateTransformer = coordinateTransformer;
    this.dataProcessor = dataProcessor;
  }

  async calculateArrowPoint(
    pictographData: PictographData,
    motionData: MotionData
  ): Promise<[number, number, number]> {
    /**
     * Calculate arrow position asynchronously with full service coordination.
     */
    try {
      // STEP 1: Extract or use provided motion data
      // ✅ FIXED: Pass color directly since ArrowPlacementData no longer has color
      const motion = motionData;
      if (!motion) {
        console.warn(
          "🚫 ArrowPositioningOrchestrator: No motion data provided, using scene center"
        );
        const center = this.coordinateSystem.getSceneCenter();
        return [center.x, center.y, 0];
      }

      // STEP 2: Calculate location and initial position
      const location = this.locationCalculator.calculateLocation(
        motion,
        pictographData
      );

      const initialPosition = this.coordinateSystem.getInitialPosition(
        motion,
        location
      );

      const validPosition =
        this.dataProcessor.ensureValidPosition(initialPosition);

      // STEP 3: Calculate rotation
      const rotation = this.rotationCalculator.calculateRotation(
        motion,
        location
      );

      // STEP 4: Calculate adjustment using sophisticated service
      const adjustment = await this.adjustmentCalculator.calculateAdjustment(
        pictographData,
        motion,
        location,
        motion.color // ✅ FIXED: Use color from MotionData as optional parameter
      );
      const [adjustmentX, adjustmentY] =
        this.dataProcessor.extractAdjustmentValues(adjustment);

      // STEP 5: Apply rotation transformation to adjustment coordinates
      const [transformedAdjustmentX, transformedAdjustmentY] =
        this.coordinateTransformer.transformAdjustmentByRotation(
          adjustmentX,
          adjustmentY,
          rotation
        );

      // STEP 6: Combine all positioning calculations
      const finalX = validPosition.x + transformedAdjustmentX;
      const finalY = validPosition.y + transformedAdjustmentY;

      return [finalX, finalY, rotation];
    } catch (error) {
      console.error("Arrow positioning calculation failed:", error);
      const center = this.coordinateSystem.getSceneCenter();
      return [center.x, center.y, 0];
    }
  }

  async updateArrowPosition(
    pictographData: PictographData,
    color: string,
    _motionData?: MotionData
  ): Promise<PictographData> {
    /**
     * Update arrow position in pictograph data.
     */
    try {
      const motionData =
        pictographData.motions?.[color as keyof typeof pictographData.motions];
      const arrowData = motionData?.arrowPlacementData;
      if (!arrowData) {
        console.warn(`No arrow data found for color: ${color}`);
        return pictographData;
      }

      const [x, y, rotation] = await this.calculateArrowPoint(
        pictographData,
        motionData
      );

      // CRITICAL: Also calculate mirroring for this arrow
      const shouldMirror = this.shouldMirrorArrow(arrowData, pictographData);

      const updates: Partial<ArrowPlacementData> = {
        positionX: x,
        positionY: y,
        rotationAngle: rotation,
        svgMirrored: shouldMirror, // ✅ FIXED: Use correct property name
      };

      // 🚨 CRITICAL FIX: Pass the calculated arrow location to persist in pictograph data
      const motionUpdates = {
        arrowLocation: location, // Use the calculated location from above
      } as unknown as Partial<MotionData>;

      return this.dataProcessor.updateArrowInPictograph(
        pictographData,
        color,
        updates,
        motionUpdates
      );
    } catch (error) {
      console.error("Arrow position update failed:", error);
      return pictographData;
    }
  }

  async calculateAllArrowPoints(
    pictographData: PictographData
  ): Promise<PictographData> {
    /**
     * Calculate positions for all arrows in the pictograph.
     */
    try {
      if (!pictographData.motions) {
        return pictographData;
      }

      let updatedPictograph = { ...pictographData };

      for (const color of Object.keys(pictographData.motions)) {
        const motionData =
          pictographData.motions[color as keyof typeof pictographData.motions];
        const arrowData = motionData?.arrowPlacementData;
        if (arrowData && motionData) {
          // Calculate location first to store it properly
          const calculatedLocation = this.locationCalculator.calculateLocation(
            motionData,
            updatedPictograph
          );

          const [x, y, rotation] = await this.calculateArrowPoint(
            updatedPictograph,
            motionData
          );

          // CRITICAL: Also calculate mirroring for this arrow
          const currentMotionData =
            updatedPictograph.motions?.[
              color as keyof typeof updatedPictograph.motions
            ];
          const shouldMirror = this.shouldMirrorArrow(
            arrowData,
            updatedPictograph,
            currentMotionData
          );

          const updates: Partial<ArrowPlacementData> = {
            positionX: x,
            positionY: y,
            rotationAngle: rotation,
            svgMirrored: shouldMirror, // ✅ FIXED: Use correct property name
          };

          // 🚨 CRITICAL FIX: Pass the calculated arrow location to persist in pictograph data
          const motionUpdates = {
            arrowLocation: calculatedLocation, // Use the calculated location from above
          } as unknown as Partial<MotionData>;

          updatedPictograph = this.dataProcessor.updateArrowInPictograph(
            updatedPictograph,
            color,
            updates,
            motionUpdates
          );
        }
      }

      return updatedPictograph;
    } catch (error) {
      console.error("Failed to calculate all arrow positions:", error);
      return pictographData;
    }
  }

  shouldMirrorArrow(
    _arrowData: ArrowPlacementData,
    pictographData?: PictographData,
    motionData?: MotionData
  ): boolean {
    /**
     * Determine if arrow should be mirrored based on motion type and prop rotation direction.
     *
     * Mirror conditions (matching desktop logic):
     * - Anti motion + clockwise → Mirror = True
     * - Anti motion + counterclockwise → Mirror = False
     * - Pro motion + clockwise → Mirror = False
     * - Pro motion + counterclockwise → Mirror = True
     * - Other motions follow "pro" rules
     */

    // Get motion data for this arrow's color
    if (!pictographData?.motions || !motionData) {
      return false;
    }

    const motionType = motionData.motionType?.toLowerCase();
    const propRotDir = motionData.rotationDirection?.toLowerCase();

    if (!motionType || !propRotDir) {
      return false;
    }

    // Mirror conditions matching desktop implementation
    const mirrorConditions = {
      anti: { cw: true, ccw: false },
      other: { cw: false, ccw: true },
    };

    // Use "anti" conditions for anti motion, "other" for everything else (pro, static, dash, float)
    const conditionKey = motionType === "anti" ? "anti" : "other";
    const shouldMirror =
      mirrorConditions[conditionKey][
        propRotDir as keyof typeof mirrorConditions.anti
      ] ?? false;

    return shouldMirror;
  }

  applyMirrorTransform(
    arrowItem: HTMLElement | SVGElement,
    shouldMirror: boolean
  ): void {
    /**
     * Apply mirror transformation to arrow graphics item.
     */
    if (shouldMirror) {
      arrowItem.style.transform = `${arrowItem.style.transform || ""} scaleX(-1)`;
    } else {
      // Remove mirror transformation
      const transform = arrowItem.style.transform || "";
      arrowItem.style.transform = transform
        .replace(/scaleX\(-1\)\s*/g, "")
        .trim();
    }
  }
}
